java/core/src/main/java/com/redhat/rhn/manager/action/ActionManager.java

```diff
@@ -1086,68 +1086,14 @@ public static PackageAction schedulePackageRunTransaction(User scheduler, Server
-                params.put("a", pm.getSystem().getArch() != null ?
-                        pm.getSystem().getArch() : "");
-                m.executeUpdate(params);
+                handleKeyThisOnly(params, pm, m);
             }
             else if (pm.getComparisonAsInt() == PackageMetadata.KEY_OTHER_ONLY) {
-
-                if (log.isDebugEnabled()) {
-                    log.debug("compare returned [KEY_OTHER_ONLY]; installing package to system: {}-{}",
-                            pm.getName(), pm.getOtherEvr());
-                }
-
-                params.put("operation", ActionFactory.TXN_OPERATION_INSERT);
-                params.put("n", pm.getName());
-                params.put("v", pm.getOther().getVersion());
-                params.put("r", pm.getOther().getRelease());
-                String epoch = pm.getOther().getEpoch();
-                params.put("e", StringUtils.isEmpty(epoch) ? null : epoch);
-                params.put("a", pm.getOther().getArch() != null ?
-                        pm.getOther().getArch() : "");
-                m.executeUpdate(params);
-
+                handleKeyOtherOnly(params, pm, m);
             }
             else if (pm.getComparisonAsInt() == PackageMetadata.KEY_THIS_NEWER ||
                     pm.getComparisonAsInt() == PackageMetadata.KEY_OTHER_NEWER) {
-
-                if (log.isDebugEnabled()) {
-                    log.debug("compare returned [KEY_THIS_NEWER OR KEY_OTHER_NEWER]; deleting package [{}-{}] " +
-                                    "from system installing package [{}-{}] to system",
-                            pm.getName(), pm.getSystemEvr(), pm.getName(), pm.getOther().getEvr());
-                }
-
-                String epoch;
-                if (isPackageRemovable(pm.getName())) {
-                    params.put("operation", ActionFactory.TXN_OPERATION_DELETE);
-                    params.put("n", pm.getName());
-                    params.put("v", pm.getSystem().getVersion());
-                    params.put("r", pm.getSystem().getRelease());
-                    epoch = pm.getSystem().getEpoch();
-                    params.put("e", StringUtils.isEmpty(epoch) ? null : epoch);
-                    params.put("a", pm.getSystem().getArch() != null ? pm.getOther().getArch() : "");
-                    m.executeUpdate(params);
-                }
-
-                params.put("operation", ActionFactory.TXN_OPERATION_INSERT);
-                params.put("n", pm.getName());
-                params.put("v", pm.getOther().getVersion());
-                params.put("r", pm.getOther().getRelease());
-                epoch = pm.getOther().getEpoch();
-                params.put("e", StringUtils.isEmpty(epoch) ? null : epoch);
-                params.put("a", pm.getOther().getArch() != null ? pm.getOther().getArch() : "");
-                m.executeUpdate(params);
+                handleKeyThisOrOtherNewer(params, pm, m);
             }
```

Comment:
This could be a actually a good place for a oop solution (naming can be improved :slightly_smiling_face:):

```java
            RunTransactionHandler handler = RunTransactionHandlerFactory.crateHandler(pm.getComparisonAsInt());
            handler.process(pm);
```

You can have three different implementation instead of three methods and the `WriteMode` could be created internally in the implementations instead of passed around.